### PR TITLE
Simulation timesteps per hr, weather resource file processes subhourly epws, run period

### PR DESCRIPTION
**Pull request by joseph-robertson**

### CODE REVIEW

**Comment by joseph-robertson** [comment]
@shorowit, how does this make you feel?

**Comment by joseph-robertson** [comment]
@shorowit did you finish review of this?

File: measures/ResidentialLocation/measure.rb [hunk 1]
```diff
@@ -51,6 +51,17 @@ def arguments(model)
     arg.setDefaultValue("October 26")
     args << arg
 
+    #make an argument for the simulation timesteps per hour
+    arg_chs = OpenStudio::StringVector.new
+    arg_chs << "60"
+    arg_chs << "6"
+    arg_chs << "4"
+    arg_chs << "2"
```
Comment (shorowit) [inline measures/ResidentialLocation/measure.rb]: Either need additional choices (30, 20, 15, 12, 10, 5, 3, 1) or ask for a double instead and make sure it divides evenly into 60.

Comment (shorowit) [reply measures/ResidentialLocation/measure.rb]: That said, it's super gross to have a "timesteps per hour" argument for the location measure. It should really be pulled out into a separate "Simulation Controls" measure.

File: measures/ResidentialLocation/measure.rb [hunk 2]
```diff
@@ -68,14 +79,15 @@ def run(model, runner, user_arguments)
     weather_file_name = runner.getStringArgumentValue("weather_file_name", user_arguments)
     dst_start_date = runner.getStringArgumentValue("dst_start_date", user_arguments)
     dst_end_date = runner.getStringArgumentValue("dst_end_date", user_arguments)
+    timesteps_per_hr = runner.getStringArgumentValue("timesteps_per_hr",user_arguments)
```
[the remark this answers is not in the document]
Comment (shorowit) [reply measures/ResidentialLocation/measure.rb]: Yes, that sounds like a good idea.

File: measures/TimeseriesCSVExport/measure.rb [hunk 3]
```diff
@@ -248,8 +248,12 @@ def run(runner, user_arguments)
       end
     end
 
-    # Get the timestamps for actual year epw file
+    # Get the run period
+    run_period = model.getRunPeriod
+
+    # Get the timestamps for actual year epw file, and the number of intervals per hour
     actual_timestamps = WeatherProcess.actual_timestamps(model, runner, File.dirname(__FILE__))
+    records_per_hour = WeatherProcess.records_per_hour(model, runner, File.dirname(__FILE__))
```
Comment (shorowit) [inline measures/TimeseriesCSVExport/measure.rb]: Why can't you get this information from weather.header? And then delete the above the method from weather.rb.

File: measures/ResidentialLocation/measure.rb (hunk 1 above)
Comment (jmaguire1) [reply measures/ResidentialLocation/measure.rb]: We might consider combining this with a future run period measure as a more general "Simulation Controls" measure: https://trello.com/c/YzbgCOcu/98-run-period-measure

File: README.md [hunk 4]
```diff
@@ -100,6 +100,7 @@ The order in which these measures are called is important. For example, the Wind
 |9. Airflow|1. Airflow|Location, HVAC Equipment, Clothes Dryer|
 |10. Sizing|1. HVAC Sizing|(lots of measures...)|
 |11. Photovoltaics|1. Photovoltaics||
+|12. Simulation Controls|1. Simulation Controls||
```
Comment (shorowit) [inline README.md]: Seems like this should be first, not last. We were previously calling `Simulation.apply` at the very beginning of the workflow (in the Location measure).

File: measures/SimulationControls/measure.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+      return false
+    end
+
+    {begin_month=>begin_day_of_month, end_month=>end_day_of_month}.each_with_index do |(month, day), i|
+      day_of_month_valid = false
+      case month
+      when 1
+        day_of_month_valid = (1..31).to_a.include? day
+      when 2
+        day_of_month_valid = (1..29).to_a.include? day
```
Comment (shorowit) [inline measures/SimulationControls/measure.rb]: So much code. Why not remove the `case` statement and just reference `Constants.MonthNumDays`?

File: measures/SimulationControls/measure.rb [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    success = Simulation.apply(model, runner, timesteps_per_hr)
+    return false if not success
+
+    runner.registerInfo("Set the simulation timesteps per hour to #{timesteps_per_hr}.")
+
+    run_period = model.getRunPeriod
+    run_period.setBeginMonth(begin_month)
+    run_period.setBeginDayOfMonth(begin_day_of_month)
+    run_period.setEndMonth(end_month)
+    run_period.setEndDayOfMonth(end_day_of_month)
```
Comment (shorowit) [inline measures/SimulationControls/measure.rb]: I suggest moving this code into the simulation.rb apply method and pass these 4 inputs.

File: measures/SimulationControls/measure.rb [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    if 60 % timesteps_per_hr != 0
+      runner.registerError("User-entered #{timesteps_per_hr} timesteps per hour does not divide evenly into 60.")
+      return false
+    end
+
+    timesteps_per_hr = timesteps_per_hr.to_i
+    begin_month = begin_month.to_i
+    begin_day_of_month = begin_day_of_month.to_i
+    end_month = end_month.to_i
+    end_day_of_month = end_day_of_month.to_i
```
Comment (shorowit) [inline measures/SimulationControls/measure.rb]: These should all be integer arguments then, no?

**Comment by joseph-robertson** [comment]
@shorowit, this should be ready for your review.

**Pull request merged**